### PR TITLE
WIP: Add an option for maximizing lines

### LINKED ISSUE #2437: Bring back longer lines
Since the change to the "heuristics" model of line length limits, I have been unhappy with the line formatting. I can set my max_width to 100, but many lines that are far below that limit get broken up -- often at something like a 1:4 ratio (!) which leads to a huge increase in vertical size of my code.

@nrc commented that "we'll have to see that the demand from users was there" here: https://github.com/rust-lang-nursery/rustfmt/issues/1984#issuecomment-353467262

so consider this my input into the demand for more flexible configuration.

### PR DESCRIPTION
**Pull request by radix**
#Fixes #2437 

This makes `use_small_heuristics` into a tri-state option with an explicit enum.

- `Minimum` -- equivalent to old `use_small_heuristics: false`
- `Medium` -- equivalent to old `use_small_heuristics: true` (default)
- `Maximum` -- new behavior: set width heuristics equal to max_width.

TODO:
- [x] Update existing tests
- [x] document the new behavior (a better description would be greatly appreciated)
- [x] Add new tests (just a copy of chains.rs for now, should I do something more specific?)
- [ ] `use_small_heuristics` obviously isn't a good name for this any more
- [ ] determine if this is even what other people want

Feedback appreciated!

### CODE REVIEW

**Comment by radix** [comment]
here's a great example of why I wanted this change, from my real codebase.

This is what rustfmt generated previously:


```rust
      TransferItem {
        from,
        to,
        item_id,
        count,
      } => self.change_with(GameLog::TransferItem {
        from,
        to,
        item_id,
        count,
      }),
      RemoveItem {
        owner,
        item_id,
        count,
      } => self.change_with(GameLog::RemoveItem {
        owner,
        item_id,
        count,
      }),
```

and now, with my changes, it's:

```rust
      TransferItem { from, to, item_id, count } => {
        self.change_with(GameLog::TransferItem { from, to, item_id, count })
      }
      RemoveItem { owner, item_id, count } => {
        self.change_with(GameLog::RemoveItem { owner, item_id, count })
      }
```

which, you'll notice, even fits within 80 lines.

File: rustfmt-config/src/lib.rs [hunk 1]
```diff
@@ -48,8 +48,7 @@ create_config! {
     tab_spaces: usize, 4, true, "Number of spaces per tab";
     newline_style: NewlineStyle, NewlineStyle::Unix, true, "Unix or Windows line endings";
     indent_style: IndentStyle, IndentStyle::Block, false, "How do we indent expressions or items.";
-    use_small_heuristics: bool, true, false, "Whether to use different formatting for items and\
-        expressions if they satisfy a heuristic notion of 'small'.";
+    use_small_heuristics: LineWidth, LineWidth::Medium, false, "The general width of lines.";
```
Comment (nrc) [inline rustfmt-config/src/lib.rs]: I think it is better to talk about how aggressively we put items across multiple lines, rather than 'general width of lines', which seems to invite confusion with `max_width`. So, rather than LineWidth::Minimum, we talk about making items as vertically formatted as possible, whereas Maximum is about making items maximally horizontal. (Or mutli-line vs single-line, if you prefer)

File: rustfmt-config/src/options.rs [hunk 2]
```diff
@@ -235,6 +244,17 @@ impl WidthHeuristics {
...
     }
+
+    pub fn maximum(max_width: usize) -> WidthHeuristics {
+        WidthHeuristics {
+            fn_call_width: max_width,
+            struct_lit_width: max_width,
+            struct_variant_width: max_width,
+            array_width: max_width,
+            chain_width: max_width,
+            single_line_if_else_max_width: max_width,
```
Comment (nrc) [inline rustfmt-config/src/options.rs]: I would use `usize::max_value()` rather than `max_width` means one less arg, plus no ordering constraint on initialisation.

File: rustfmt-config/src/options.rs [hunk 3]
```diff
@@ -189,6 +189,15 @@ configuration_option_enum! { Color:
     Auto,
 }
 
+configuration_option_enum! { LineWidth:
+    // Always format lines as short as possible (?)
+    Minimum,
```
Comment (nrc) [inline rustfmt-config/src/options.rs]: The `null` width heuristic doesn't make lines as short as possible, it tries to layout stuff in its 'default' layout without taking into consideration the width of the sub-item. I.e., it will make function calls as wide as possible, but will never put if/else on one line (since multi-line is the default).

File: rustfmt-config/src/options.rs (hunk 2 above)
Comment (radix) [reply rustfmt-config/src/options.rs]: fwiw, the `scaled()` constructor also takes max_width, but I can still change this np.

File: rustfmt-config/src/lib.rs (hunk 1 above)
Comment (radix) [reply rustfmt-config/src/lib.rs]: This is still the biggest thing holding me back from finishing up this PR. The line length heuristics are confusing. I don't think "more horizontal" vs "more vertical" even really describes the options.

#### `use_small_heuristics: true`: (default)

- function calls: 60%
- struct literals: 18%
- struct variants: 35%
- arrays: 60%
- chains: 60%
- if/else: 50%

#### `use_small_heuristics: false`

- function calls: 100%
- struct literals: 0%
- struct variants: 0%
- arrays: 100%
- chains: 100%
- if/else: 0%

This means that enabling heuristics leads to code being sometimes more vertical and sometimes less vertical. This makes it really hard for me to describe what `use_small_heuristics: false` does in plain English, which also makes it hard to come up with a non-boolean name for what it does :(

Comment (nrc) [reply rustfmt-config/src/lib.rs]: Hmm, yeah I know what you mean. The logic is that with `use_small_heuristics` false, we use the same formatting for small things as we do for large things, whereas when true we do things differently depending on the size - we've never tried to frame it as making lines longer or shorted, just 'better' (easier to read, prettier, etc). I guess the trouble with describing this is that even without `use_small_heuristics`, the set of things which is put on one line, rather than multiple lines is somewhat arbitrary.

Perhaps values like `None` (today's false), `Default` (today's true), and `Max` (for the new stuff) is OK. They're not very self-explanatory, but I don't think we'll do better.

File: rustfmt-config/src/options.rs (hunk 2 above)
Comment (nrc) [reply rustfmt-config/src/options.rs]: `scaled` needs it though since it uses the max width as input (we could probably use `max_value` there for the 100% ones).

**Comment by radix** [comment]
Hi, just an update on this PR to state that I *do* still intend to finish this off, I have just been busy with other things lately. I'm closing this PR to unclutter the PR list for the maintainers, but I will re-open it once I've addressed the comments.

Thanks @nrc for being so helpful with the reviews!
